src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    private class GenericMenuEntry {
+        private String id;
+        private String description;
+        private String parent;
+        private String functionId;
+
+        public GenericMenuEntry(FOLDER_Type folder) {
+            this.id = folder.getEXTMENUCODE();
+            this.description = folder.getFOLDERID().getFOLDERDESCRIPTION();
+            this.parent = folder.getEXTMENUPARENT();
+            this.functionId = folder.getFOLDERID().getFOLDERCODE();
+        }
+
+        public GenericMenuEntry(MENU_Type menu) {
+            this.id = menu.getEXTMENUCODE();
+            this.description = menu.getMENUID().getMENUDESCRIPTION();
+            this.parent = menu.getEXTMENUPARENT();
+            this.functionId = menu.getMENUID().getMENUCODE();
+        }
+
+        public GenericMenuEntry(FUNCTION_Type function) {
+            this.id = function.getEXTMENUCODE();
+            this.description = function.getFUNCTIONID().getFUNCTIONDESCRIPTION();
+            this.parent = function.getEXTMENUPARENT();
+            this.functionId = function.getFUNCTIONID().getFUNCTIONCODE();
+        }
+
+        public String getId() {
+            return this.id;
+        }
+
+        public String getDescription() {
+            return this.description;
+        }
+
+        public String getFunctionId() {
+            return this.functionId;
+        }
+
+        public Object getParent() { return this.parent; }
+    }
+
+    private ExtMenusHierarchy getExtMenuHierarchy(InforContext context, MenuSpecification node) throws InforException {
+        MP6005_GetExtMenusHierarchy_001 getExtMenusHierarchy = new MP6005_GetExtMenusHierarchy_001();
+        getExtMenusHierarchy.setUSERGROUPID(new USERGROUPID_Type());
+        getExtMenusHierarchy.getUSERGROUPID().setUSERGROUPCODE(node.userGroup);
+
+        ExtMenusHierarchy result =
+                tools.performInforOperation(context, inforws::getExtMenusHierarchyOp, getExtMenusHierarchy)
+                        .getResultData().getExtMenusHierarchy();
+
+        return result;
+    }
+
+    private List<GenericMenuEntry> getExtMenuHierarchyAsList(InforContext context, MenuSpecification node) throws InforException {
+        ExtMenusHierarchy result = this.getExtMenuHierarchy(context, node);
+
+        List<GenericMenuEntry> menuEntries = new ArrayList<>();
+        for(MENU_Type menu : result.getMENU()) {
+            menuEntries.add(new GenericMenuEntry(menu));
+            menu.getFOLDER().stream().forEach(childFolder -> this.addFolderToEntries(menuEntries, childFolder));
+            menu.getFUNCTION().stream().forEach(childFunction -> this.addFunctionToEntries(menuEntries, childFunction));
+        }
+
+        return menuEntries;
+    }
+
+    private String decideMenuType(String menuCode, String[] path) {
+        String menuType;
+        if (menuCode == null || menuCode.isEmpty()) {
+            if (path.length <= 1) {
+                menuType = "M"; // Item is main menu
```

Comment:
Using an enum instead of a String would both be more readable and prevent errors from usage

Reply:
Used an enum now following another ws-hub-core enum/dictionary. (I would personally use a class with public static final strings, since it's less dirty).